USF-1329: Checkout functions (#52)
commerce-docs/microsite-commerce-storefront: src/content/docs/dropins/checkout/checkout-functions.mdx

---
title: Checkout functions
description: Learn about the API functions provided by the Checkout dropin.
---

import Aside from '@components/Aside.astro';

This topic provides the details and instructions you need to use the functions provided by the Checkout dropin.

## estimateShippingMethods

Fetches and displays available shipping methods based on a customer's address information. It can take a combination of fields as `criteria`: country, region, region identifier, and postal code.

### Returns

> Uses a utility function to send an [`estimateShippingMethodsMutation`](https://developer.adobe.com/commerce/webapi/graphql-api/index.html#mutation-estimateShippingMethods) request to the Adobe Commerce GraphQL API.

Reviewer: This line belongs in the overview section. The next few functions have the cross reference there.

const input = {
  criteria: {
    country_code: 'US',
    region_name: 'FL',
  },
};
const estimateShippingMethods = estimateShippingMethods(input);

// By country code and region ID
const input = {
  criteria: {
    country_code: 'US',
    region_id: 18,
  },
};
const estimateShippingMethods = estimateShippingMethods(input);

// By country code and postal code
const input = {
  criteria: {
    country_code: 'US',
    zip: '80000',
  },
};
const estimateShippingMethods = estimateShippingMethods(input);

// By country code, region and postal code
const input = {
  criteria: {
    country_code: 'US',
    region_name: 'FL',
    zip: '80000',
  },
};
const estimateShippingMethods = estimateShippingMethods(input);

// By country code, region, region id and postal code
// This scenario is specific for those regions which have more than a region ID
// associated (i.e. Armed Forces region 'AE' with some regions IDs); if the region has
// just a region ID,
const input = {
  criteria: {
    country_code: 'US',
    region_name: 'FL',
    region_id: 18,
    zip: '80000',
  },
};
const estimateShippingMethods = estimateShippingMethods(input);
```

## fetchAddressFormFields

> Retrieves a list of EAV attributes associated with customer register and customer address frontend forms from the Adobe Commerce GraphQL API using the [`attributesForm`](https://developer.adobe.com/commerce/webapi/graphql/schema/attributes/queries/attributes-form/) query.

Reviewer: ```suggestion
Uses the [`attributesForm`](https://developer.adobe.com/commerce/webapi/graphql/schema/attributes/queries/attributes-form/) query to retrieve a list of EAV attributes associated with customer and customer address frontend forms.
```

If you choose to keep the sentence structure as is, note that I deleted "register". Customers can edit and register their settings.

### Returns

Returns a promise that resolves to the transformed data fetched from the API. If the API call fails, it returns an empty array.

```ts
transformAddressFormFields
```

### Usage

See the following example for usage details:

```ts
import { fetchAddressFormFields } from '@/checkout/api/fetchAddressFormFields';

const addressFormFields = fetchAddressFormFields();
```

## getCart

Fetches the cart details for either a [guest user](https://developer.adobe.com/commerce/webapi/graphql-api/index.html#query-customerCart) or an [authenticated customer](https://developer.adobe.com/commerce/webapi/graphql-api/index.html#query-cart) from the Adobe Commerce GraphQL API.

### Returns

Returns a promise that resolves to the transformed cart data fetched from the API.

```ts
transformCart
```

### Usage

See the following example for usage details:

```ts
import { getCart } from '@/checkout/api/getCart';

getCart();
```

## getCheckoutData

> Fetches and prepares the necessary data for the checkout process. It uses the `getCart` and `getCustomer` functions to fetch the cart and customer data from the Adobe Commerce GraphQL API.

Reviewer: Add a link to the `getCart` and `getCustomer` functions. Since this function doesn't retrieve GraphQL directly, I think I'd leave that out.

### Returns

The function returns a promise that resolves to the transformed data fetched from the API. If the API call fails, it returns an empty array.

```ts
transformCountries
```

### Usage

See the following example for usage details:

```ts
import { getCountries } from '@/checkout/api/getCountries';

getCountries();
```

## getCustomer

Fetches customer details for an authenticated user from the Adobe Commerce GraphQL API using the [`customer`](https://developer.adobe.com/commerce/webapi/graphql/schema/customer/queries/customer/) query.

### Returns

The function returns a promise that resolves to the transformed customer data fetched from the API or `undefined` if the user is not authenticated.

The `Customer` object is returned:

```ts
type CustomerAddress = {
  id: string;
  firstName: string;
  lastName: string;
  company?: string;
  street: string[];
  city: string;
  postCode?: string;
  vatId?: string;
  telephone?: string;
  region?: Region;
  country: Country;
  customAttributes: CustomAttribute[];
};

> type Customer = {

Reviewer: `Customer` should be listed first, IMHO.

If the API call fails, it returns the default store configuration settings (`STORE_CONFIG_DEFAULTS`).

```ts
transformStoreConfig
```

### Usage

See the following example for usage details:

```ts
import { storeConfig } from '@/checkout/api/store/storeConfig';

storeConfig();
```

## initializeCheckout

Initializes the checkout process by fetching necessary configuration and data from various Adobe Commerce GraphQL APIs using the following functions:

- `fetchAddressFormFields`
- `getCountries`
- `getStoreConfig`

### Usage

See the following example for usage details:

```typescript
import {
  fetchAddressFormFields,
  getCountries,
  getStoreConfig,
} from '@/checkout/api';

initializeCheckout();
```

### Returns

The function returns a promise that resolves to an array containing the results of the three API calls.

```ts
Promise.all([getStoreConfig(), getCountries(), fetchAddressFormFields()]);
```

## isEmailAvailable

Checks whether the specified email has already been used to create a customer account using the [`isEmailAvailable`](https://developer.adobe.com/commerce/webapi/graphql/schema/customer/queries/is-email-available/) Adobe Commerce GraphQL API query.

> A value of `true` indicates that the email address is available and the customer can use the email address to create an account.

Reviewer: As of 2.4.7, the default value is always true, for security reasons. I don't think it's necessary to explain why here, but maybe say something like "refer to the query documentation for implementation details."

### Returns

A promise that resolves to an `EmailAvailability` object indicating whether the email is available.

```ts
type EmailAvailability = boolean;
```

### Usage

See the following example for usage details:

```ts
import { isEmailAvailable } from '@/checkout/api/isEmailAvailable';

isEmailAvailable('test@example.com');
```

## placeOrder

> Handles the process of placing an order in the checkout system using the [`placeOrder`](https://developer.adobe.com/commerce/webapi/graphql/schema/cart/mutations/place-order/) Adobe Commerce GraphQL API mutation.

Reviewer 2: Also you could mention this API is publishing the `'place-order'` [Adobe Commerce Event](https://github.com/adobe/commerce-events/blob/main/examples/events/place-order.md) within the [order](https://github.com/adobe/commerce-events/blob/main/packages/storefront-events-sdk/src/types/schemas/order.ts) and [shoppingCart](https://github.com/adobe/commerce-events/blob/main/packages/storefront-events-sdk/src/types/schemas/shoppingCart.ts) contexts to the ACDL ([Adobe Client Data Layer](https://github.com/adobe/adobe-client-data-layer)).

### Returns

The transformed order data fetched from the API.

```ts
transformOrder
```

### Usage

See the following example for usage details:

```ts
import { placeOrder } from '@/checkout/api/placeOrder';

> placeOrder('4L7E1H0muABuggqWxbzBCBWoeG0rNxDh');

Reviewer 2: `placeOrder()` API is without params (I guess mdx file was not updated correctly).

```

## redirect

A function that redirects the user to a URL. It takes a single parameter (`url`) of type `string`, which represents the URL to redirect to.

### Usage

See the following example for usage details:

```ts
import { redirect } from '@/checkout/api/redirect';

redirect('https://example.com');
```

## resetCheckout

Resets the checkout process to its initial state by clearing the cart data.

### Usage

See the following example for usage details:

```ts
import { state } from '@/checkout/lib/state';

resetCheckout();
```

## resetCustomer

Clears the current customer information and resets the customer state to its inital value in the checkout process.

### Usage

See the following example for usage details:

```typescript
import { state } from '@/checkout/lib/state';

resetCustomer();
```

## setBillingAddress

Sets the billing address for a specific cart using the [`setBillingAddressOnCart`](https://developer.adobe.com/commerce/webapi/graphql/schema/cart/mutations/set-billing-address/) Adobe Commerce GraphQL mutation.

> It takes the following parameters:

Reviewer 2: We refactored the input parameters to keep compatibility with the mutation as follow:
`setBillingAddress({input: BillingAddressInput})`
...where 
```
BillingAddressInput {
  address?: AddressInput;
  customer_address_id?: number;
  same_as_shipping?: boolean;
  use_for_shipping?: boolean;
}
```
Then an example could be:
```
setBillingAddress({
  input: {
    address: {
      firstname: 'John',
      lastname: 'Doe',
      street: ['123 Main St', 'Apt 1'],
      city: 'New York',
      postcode: '10001',
      country_code: 'US',
      telephone: '555-555-5555',
    },
    same_as_shipping: false,
  }
});
```
I guess the mdx file is not up-to-date.

- `address`: An object that contains the billing address details.
- `customer_address_id`: The ID of the customer address.
- `same_as_shipping`: A boolean value that indicates whether the billing address is the same as the shipping address.
- `use_for_shipping`: A boolean value that indicates whether the billing address should be used as the shipping address.

### Returns

The function returns a promise that resolves to the transformed cart data fetched from the API.

```ts
transformCart
```

### Usage

To set the billing address as the same as the shipping address:

```ts
import { setBillingAddress } from '@/checkout/api/setBillingAddress';

setBillingAddress({
  same_as_shipping: true,
});
```

To set an specific billing address:

```ts
import { setBillingAddress } from '@/checkout/api/setBillingAddress';
setBillingAddress({
  address: {
    firstname: 'John',
    lastname: 'Doe',
    street: ['123 Main St', 'Apt 1'],
    city: 'New York',

>     state: 'NY',

Reviewer: `state` is not part of the schema. It should be `region` or maybe a `region` object

    postal_code: '10001',
    country: 'US',
    telephone: '555-555-5555',
  },
  same_as_shipping: false,
});
```

## setGuestEmailOnCart

Sets the email address for a guest user on the cart using the [`setGuestEmailOnCart`](https://developer.adobe.com/commerce/webapi/graphql/schema/cart/mutations/set-guest-email/) Adobe Commerce GraphQL API mutation.

It takes the following parameters:

- `email`: The email address of the guest.

> - `cart_id`: The unique ID of the cart object.

Reviewer 2: The only param is `email` (cart_id is got from an internal state).

### Returns 

The function returns a promise that resolves to the transformed cart data fetched from the API.

```ts
transformCart
```

### Usage

See the following example for usage details:

```ts
import { setGuestEmailOnCart } from '@/checkout/api/setGuestEmailOnCart';

> setGuestEmailOnCart({

Reviewer 2: The example should be:
`setGuestEmailOnCart('test@example.com')`

It takes the following parameter: `paymentMethod`.

### Returns

The function returns a promise that resolves to the transformed cart data fetched from the API.

```ts
transformCart
```

### Usage

See the following example for usage details:

```ts
import { setPaymentMethod } from '@/checkout/api/setPaymentMethod';

setPaymentMethod('payment-method-code');
```

## setShippingAddress

Sets the shipping address for a specific cart using the [`setShippingAddressesOnCart`](https://developer.adobe.com/commerce/webapi/graphql/schema/cart/mutations/set-shipping-address/) Adobe Commerce GraphQL API mutation.

It takes the following parameter: `address`.

### Returns

The function returns a promise that resolves to the transformed cart data fetched from the API.

```ts
transformCart
```

### Usage

See the following example for usage details:

```ts
import { setShippingAddress } from '@/checkout/api/setShippingAddress';

setShippingAddress({
  address: {
    city: 'San Francisco',
    country_code: 'US',
    firstname: 'John',
    lastname: 'Doe',
    postcode: '94103',

>     region: {

Reviewer: `region` looks correct here
